stumpy/core.py

```diff
@@ -2254,15 +2349,108 @@ def rolling_isconstant(a, w):
     w : numpy.ndarray
         The rolling window size
 
+    custom : np.ndarray or function, default None
+        A boolean array that indicates whether a subsequence in `T` is constant
+        (True). Alternatively, a custom, user-defined function that returns a
+        boolean array that indicates whether a subsequence in `T` is constant
+        (True). The function must only take two arguments, `a`, a 1-D array,
+        and `w`, the window size, while additional arguments may be specified
+        by currying the user-defined function using `functools.partial`. When
+        None, this defaults to `_rolling_isconstant`.
+
     Returns
     -------
-    output : numpy.ndarray
-        Rolling window isconstant.
+    T_subseq_isconstant : numpy.ndarray
+        Rolling window isconstant
+    """
+    if custom is None:
```

Comment:
I think I can understand your mindset. You are trying to first do some sort of preprocessing, and then start computing the ouput if needed (i.e. when the input is not `np.ndarray`). This sounds great. 

I have a couple of comments:

(1) This very first line in the code, i.e. 

> isconstant_func = _rolling_isconstant

should be `T_subseq_isconstant = _rolling_isconstant`. Right? Because, if input `T_subseq_isconstant` is None, then the current code raises an error.

(2) I couldn't find an easy way to find the arguments of function with no default value. Currently, as you can see in the code, I am doing:

```
custom_args = []
        for arg_name, arg in inspect.signature(custom).parameters.items():
            if arg.default == inspect.Parameter.empty:
                custom_args.append(arg_name)
```

Is there an easier way?

Reply:
> I think I can understand your mindset. You are trying to first do some sort of preprocessing, and then start computing the ouput if needed (i.e. when the input is not np.ndarray). This sounds great.

Yes, I think one can get too "smart" by trying to fit everything into nested if/else statements. However, this becomes very hard to read/follow. Instead, what I've done is split things up into logical parts. By analogy, when you are an expert chef, you can certainly boil the water, chop the veggies while the water is boiling, and then cook the pasta at the same time and be super efficient in order to save yourself 3 minutes. However, when you're providing instructions for somebody else to read, it's better to chop the veggies first, then boil the water, and then finally cook the pasta all in separate steps as it will be easier to follow and, more importantly, easy for ANYBODY to read the recipe in the future.

> should be T_subseq_isconstant = _rolling_isconstant. Right? Because, if input T_subseq_isconstant is None, then the current code raises an error.

You're right, I missed the `T_subseq_isconstant is None` case. Perhaps, it should start with:

```
if T_subseq_isconstant is None:
    T_subseq_isconstant = _rolling_isconstant
```

> I couldn't find an easy way to find the arguments of function with no default value.

I'm not sure why you are looking "for arguments of function with no default value". What is wrong with what I proposed?

```
func_args = list(inspect.signature(T_subseq_isconstant).parameters.items())
incomp_args = set(func_args).difference({"a", "w"})
```

The goal is simply to check whether the signature of the supplied function has exactly two arguments (no more, no less) and that they are named `a` and `w`. I think my proposed code does exactly that. I'm not sure what that has to do with `default value`. Can you please elaborate?

Reply:
Using `.keys` should work for all of the following functions

```
def some_func(a, w):
    return


def some_func(a=10, w):
    return

def some_func(a=10, w=20):
    return
```

In all three cases, `list(inspect.signature(T_subseq_isconstant).parameters.keys())` would return `[a, w]`

Reply:
@seanlaw 
Unfortunately, this does not work when `functools.partial` is used. Maybe I am doing something wrong, but this is what I get:

```
import functools
import inspect

def custom_func(a, w, extra_param):
    return 

f = functools.partial(custom_func, extra_param=None)

params_name = list(inspect.signature(f).parameters.keys())
print(params_name)

# output
['a', 'w', 'extra_param']
```

Reply:
@NimaSarajpoor Thanks for the explanation and it looks like your proposal solves the problem. I think it would help to explain the functools.partial case as a comment.

This is basically the same as your solution and I'm not sure if it's better/worse:

```
func_args = [k for k, v in inspect.signature(f).parameters.items() if v.default == inspect.Parameter.empty]
``` 

I think your approach is a little more readable

Reply:
Let's make sure to add a `functools.partial` example in our unit test. This will protect us in the future if we accidentally change the argument checking logic (i.e., the `functools.partial` test would fail)

Reply:
It is already added to the unit test. In fact, that resulted in raising an error, which in turn helped me with solving it.

Reply:
Awesome! I had missed that since I had not reviewed the tests yet. You're a pro!

Reply:
Thank you for your kind words. Still having a long way to get there :) but I am pro in enjoying the journey 😀